lean4/src/putnam_2023_a6.lean

```diff
@@ -0,0 +1,22 @@
+import Mathlib
+
+open Finset
+
+abbrev putnam_2023_a6_solution : Set ℕ := sorry
+-- {n : ℕ | 0 < n}
+/--
+Alice and Bob play a game in which they take turns choosing integers from $1$ to $n$. Before any integers are chosen, Bob selects a goal of 'odd' or 'even'. On the first turn, Alice chooses one of the $n$ integers. On the second turn, Bob chooses one of the remaining integers. They continue alternately choosing one of the integers that has not yet been chosen, until the $n$th turn, which is forced and ends the game. Bob wins if the parity of $\{k : \mbox{the number $k$ was chosen on the $k$th turn}\}$ matches his goal. For which values of $n$ does Bob have a winning strategy?
+-/
+theorem putnam_2023_a6
+    (IsValidGame : List ℕ → Prop)
+    (IsValidGame_def : ∀ g, IsValidGame g ↔ g.Nodup ∧ (∀ i ∈ g, i ∈ Icc 1 g.length))
+    (parityOf : List ℕ → ZMod 2)
+    (parityOf_def : ∀ g, parityOf g = ((range g.length).filter fun i ↦ g[i]! = i + 1).card)
+    (ConformsToStrategy : List ℕ → (List ℕ → ℕ) → Prop)
+    (ConformsToStrategy_def : ∀ g s, ConformsToStrategy g s ↔
+      ∀ (i) (h : i < g.length), Odd i → g[i] = s (g.take i))
+    (IsWinningFor : ℕ → (List ℕ → ℕ) → Prop)
+    (IsWinningFor_def : ∀ n s, IsWinningFor n s ↔
+      ∃ p, ∀ g, g.length = n → IsValidGame g → ConformsToStrategy g s → parityOf g = p) :
```

Comment:
While I believe this formalization doesn't have an issue here, I am curious about your thoughts on this question: when writing a formal spec without a matching proof like we do here, should we include "safety blanket" hypotheses? This would give some extra confidence but introduce some clutter since it could be unnecessary. In this case, an example might be the extra constraint on the strategy's behavior (like `s L \notin L`), though I think this follows already.

Reply:
IMHO, I we should not include such hypotheses because:

- they might mask an error
- they might contain information which is the duty of the problem solver to deduce
- they are redundant

However, I also believe they would probably salvage the occasional misformalisation. So it's a subjective call really, dependent upon what you view as the job of these formalisations.

Reply:
I tend to agree with you. I think in the case with a problem like this one (where the natural language statement isn't ''easily'' formalizable) there is some ambiguity in what might be the asserted conditions we need to put in the spec. For example "Bob chooses one of the **remaining** integers" could directly translate to something like `s L \notin L`. If one agrees on this, then it is probably better that this appears in the formalization (abstractly, I'm not actually saying there should be changes, to be clear), otherwise one needs to infer that the condition follows from `IsValidGame g` and `ConformsToStrategy g s`. Perhaps it is really a matter of how faithful the formalization is to the natural language statement, which I agree is a subjective call.